tests/e2e/tests/pilot/pod_churn_test.go

```diff
@@ -0,0 +1,384 @@
+func newTrafficGenerator(t *testing.T, srcApp string, qps uint16) trafficGenerator {
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+
+	return &trafficGeneratorImpl{
+		srcApp:    srcApp,
+		qps:       qps,
+		forwarder: forwardToGrpcPort(t, srcApp),
+		wg:        wg,
+		results:   make(map[string]int),
+	}
+}
+
+func forwardToGrpcPort(t *testing.T, app string) kube.PortForwarder {
+	// Find the gRPC port for this service.
+	svc, err := tc.Kube.KubeAccessor.GetService(tc.Kube.Namespace, app)
+	if err != nil {
+		t.Fatal(err)
+	}
+	grpcPort := 0
+	for _, port := range svc.Spec.Ports {
+		if port.Name == "grpc" {
+			grpcPort = int(port.TargetPort.IntVal)
+		}
+	}
+	if grpcPort == 0 {
+		t.Fatalf("unable to locate gRPC port for service %s", app)
+	}
+
+	// Get the pods for the source App.
+	pods := tc.Kube.GetAppPods(primaryCluster)[app]
+	if len(pods) == 0 {
+		t.Fatalf("missing pod names for app %q from %s cluster", app, primaryCluster)
+	}
+
+	// Create a port forwarder so that we can send commands app "a" to talk to the churn app.
+	forwarder, err := tc.Kube.KubeAccessor.NewPortForwarder(&kube.PodSelectOptions{
+		PodNamespace: tc.Kube.Namespace,
+		PodName:      pods[0],
+	}, 0, uint16(grpcPort))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err := forwarder.Start(); err != nil {
+		t.Fatal(err)
+	}
+	return forwarder
+}
+
+// trafficGeneratorImpl implementation of the trafficGenerator interface.
+type trafficGeneratorImpl struct {
+	srcApp    string
+	qps       uint16
+	forwarder kube.PortForwarder
+	wg        *sync.WaitGroup
+
+	results map[string]int
+	err     error
+}
+
+// close implements the trafficGenerator interface.
+func (g *trafficGeneratorImpl) close() {
+	if g.forwarder != nil {
+		_ = g.forwarder.Close()
+		g.forwarder = nil
+	}
+}
+
+// awaitResults implements the trafficGenerator interface.
+func (g *trafficGeneratorImpl) awaitResults() (map[string]int, error) {
+	g.wg.Wait()
+
+	if g.err != nil {
+		return nil, g.err
+	}
+
+	return g.results, nil
+}
+
+// start implements the trafficGenerator interface.
+func (g *trafficGeneratorImpl) start() {
+	// Command to send traffic for 1s.
+	request := proto.ForwardEchoRequest{
+		Url:   churnAppURL,
+		Count: int32(g.qps),
+		Qps:   int32(g.qps),
+	}
+
+	// Create the results for this thread.
+	go func() {
+		defer g.wg.Done()
+
+		// Create a gRPC client to the source pod.
+		client, err := echo.NewClient(g.forwarder.Address())
+		if err != nil {
+			g.err = multierror.Append(g.err, err)
+			return
+		}
+		defer func() {
+			_ = client.Close()
+		}()
+
+		// Send traffic from the source pod to the churned pods for a period of time.
+		timer := time.NewTimer(churnTrafficDuration)
+
+		responseCount := 0
+		for {
+			select {
+			case <-timer.C:
+				if responseCount == 0 {
+					g.err = fmt.Errorf("service %s unable to communicate to churn app", g.srcApp)
+				}
+				return
+			default:
+				responses, err := client.ForwardEcho(&request)
```

Comment:
How does this make retry configuration take effect? Since there is no proxy here.